File: source/Tools/HdrUtilities/HdrUtilities.cpp
#include "HdrUtilities.h"

#include <vector>
#include <optional>

#include <Windows.h>

namespace
{
    std::optional<bool> NativeEnableHdr(bool enable)
    {
        UINT32 pathCount, modeCount;
        LONG result = GetDisplayConfigBufferSizes(QDC_ONLY_ACTIVE_PATHS, &pathCount, &modeCount);
        if (result != ERROR_SUCCESS)
        {
            return std::nullopt;
        }

        std::vector<DISPLAYCONFIG_PATH_INFO> paths(pathCount);
        std::vector<DISPLAYCONFIG_MODE_INFO> modes(modeCount);

        result = QueryDisplayConfig(QDC_ONLY_ACTIVE_PATHS, &pathCount, paths.data(), &modeCount, modes.data(), nullptr);
        if (result != ERROR_SUCCESS)
        {
            return std::nullopt;
        }

        for (DISPLAYCONFIG_MODE_INFO const& mode : modes)
        {
            if (mode.infoType == DISPLAYCONFIG_MODE_INFO_TYPE_SOURCE)
            {
                //The primary display is positioned at (0,0)
                if (mode.sourceMode.position.x == 0 && mode.sourceMode.position.y == 0)
                {
                    auto iter = std::find_if(
                        paths.cbegin(),
                        paths.cend(),
                        [&mode](DISPLAYCONFIG_PATH_INFO const& p) { return mode.id == p.sourceInfo.id; });

                    if (iter == paths.cend())
                    {
                        return std::nullopt;
                    }
                    DISPLAYCONFIG_PATH_TARGET_INFO const& targetInfo = iter->targetInfo;

                    //Retrieve current HDR state
                    DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO colorInfo = {};
                    colorInfo.header.type = DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO;
                    colorInfo.header.adapterId = targetInfo.adapterId;
                    colorInfo.header.id = targetInfo.id;
                    colorInfo.header.size = sizeof(colorInfo);
                    result = DisplayConfigGetDeviceInfo(&colorInfo.header);
                    if (result != ERROR_SUCCESS)
                    {
                        return std::nullopt;
                    }

                    bool wasHdrEnabled = colorInfo.advancedColorEnabled == 1;

                    //Enable or Disable HDR
                    if (colorInfo.advancedColorSupported)
                    {
                        DISPLAYCONFIG_SET_ADVANCED_COLOR_STATE newColorInfo = {};
                        newColorInfo.header.type = DISPLAYCONFIG_DEVICE_INFO_SET_ADVANCED_COLOR_STATE;
                        newColorInfo.header.adapterId = targetInfo.adapterId;
                        newColorInfo.header.id = targetInfo.id;
                        newColorInfo.header.size = sizeof(newColorInfo);

                        newColorInfo.enableAdvancedColor = enable ? 1 : 0;
                        DisplayConfigSetDeviceInfo(&newColorInfo.header);
                    }

                    return wasHdrEnabled;
                }
            }
        }
        return std::nullopt;
    }

    std::optional<bool> NativeIsHdrSupported()
    {
        UINT32 pathCount, modeCount;
        LONG result = GetDisplayConfigBufferSizes(QDC_ONLY_ACTIVE_PATHS, &pathCount, &modeCount);
        if (result != ERROR_SUCCESS)
        {
            return std::nullopt;
        }

        std::vector<DISPLAYCONFIG_PATH_INFO> paths(pathCount);
        std::vector<DISPLAYCONFIG_MODE_INFO> modes(modeCount);

        result = QueryDisplayConfig(QDC_ONLY_ACTIVE_PATHS, &pathCount, paths.data(), &modeCount, modes.data(), nullptr);
        if (result != ERROR_SUCCESS)
        {
            return std::nullopt;
        }

        for (DISPLAYCONFIG_MODE_INFO const& mode : modes)
        {
            if (mode.infoType == DISPLAYCONFIG_MODE_INFO_TYPE_SOURCE)
            {
                //The primary display is positioned at (0,0)
                if (mode.sourceMode.position.x == 0 && mode.sourceMode.position.y == 0)
                {
                    auto iter = std::find_if(
                        paths.cbegin(),
                        paths.cend(),
                        [&mode](DISPLAYCONFIG_PATH_INFO const& p) { return mode.id == p.sourceInfo.id; });

                    if (iter == paths.cend())
                    {
                        return std::nullopt;
                    }
                    DISPLAYCONFIG_PATH_TARGET_INFO const& targetInfo = iter->targetInfo;

                    //Retrieve current HDR state
                    DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO colorInfo = {};
                    colorInfo.header.type = DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO;
                    colorInfo.header.adapterId = targetInfo.adapterId;
                    colorInfo.header.id = targetInfo.id;
                    colorInfo.header.size = sizeof(colorInfo);
                    result = DisplayConfigGetDeviceInfo(&colorInfo.header);
                    if (result != ERROR_SUCCESS)
                    {
                        return std::nullopt;
                    }

                    return colorInfo.advancedColorSupported == 1;
                }
            }
        }
        return std::nullopt;
    }
}

System::Nullable<bool> HdrUtilities::EnableHdr(bool enable)
{
    std::optional<bool> wasEnabled = NativeEnableHdr(enable);
    return wasEnabled.has_value() ? System::Nullable<bool>(wasEnabled.value()) : System::Nullable<bool>();
}

bool HdrUtilities::IsHdrSupported()
{
    std::optional<bool> isSupported = NativeIsHdrSupported();
    return isSupported.has_value() ? isSupported.value() : false;
}

File: source/Tools/HdrUtilities/HdrUtilities.cs
using System.Linq;
using System;
using System.Runtime.InteropServices;

public class HdrUtilities
{
    /// <summary>
    /// Determines if HDR is supported on the primary display
    /// </summary>
    /// <returns>True if HDR is supported on the primary display, false if unknown or not supported</returns>
    public static bool IsHdrSupported()
    {
        DisplayConfig.DisplayConfigPathTargetInfo? targetInfo = GetPrimaryDisplayTargetInfo();
        if (!targetInfo.HasValue)
        {
            return false;
        }

        (bool? isHdrSupported, _) = GetHdrStatus(targetInfo.Value);
        return isHdrSupported.GetValueOrDefault(false);
    }

    /// <summary>
    /// Determines if HDR is enabled on the primary display
    /// </summary>
    /// <returns>True if HDR is enabled on the primary display, false if unknown or not supported</returns>
    public static bool? IsHdrEnabled()
    {
        DisplayConfig.DisplayConfigPathTargetInfo? targetInfo = GetPrimaryDisplayTargetInfo();
        if (!targetInfo.HasValue)
        {
            return false;
        }

        (_, bool? isHdrEnabled) = GetHdrStatus(targetInfo.Value);
        return isHdrEnabled;
    }

    /// <summary>
    /// Enables/disables HDR on the primary display
    /// </summary>
    /// <param name="enable">True if enabling HDR, false if disabling HDR</param>
    public static void EnableHdr(bool enable)
    {
        DisplayConfig.DisplayConfigPathTargetInfo? targetInfo = GetPrimaryDisplayTargetInfo();
        if (!targetInfo.HasValue)
        {
            return;
        }

        if (IsHdrSupported())
        {
            //Enable or Disable HDR
            var newColorInfo = new DisplayConfig.DisplayConfigSetAdvancedColorState();
            newColorInfo.header.type = DisplayConfig.DisplayConfigDeviceInfoType.SetAdvancedColorState;
            newColorInfo.header.adapterId = targetInfo.Value.adapterId;
            newColorInfo.header.id = targetInfo.Value.id;
            newColorInfo.header.size = Marshal.SizeOf<DisplayConfig.DisplayConfigSetAdvancedColorState>();
            newColorInfo.enableAdvancedColor = enable;

            DisplayConfig.DisplayConfigSetDeviceInfo(ref newColorInfo.header);
        }
    }

    private static bool IsPrimaryDisplayMode(DisplayConfig.DisplayConfigModeInfo mode)
    {
        return
            mode.infoType == DisplayConfig.DisplayConfigModeInfoType.Source &&
            mode.sourceMode.position.x == 0 &&
            mode.sourceMode.position.y == 0;
    }

    private static DisplayConfig.DisplayConfigPathTargetInfo? GetPrimaryDisplayTargetInfo()
    {
        long result = DisplayConfig.GetDisplayConfigBufferSizes(DisplayConfig.Qdc.OnlyActivePaths, out uint pathCount, out uint modeCount);
        if (result != DisplayConfig.ErrorSuccess)
        {
            return null;
        }

        var paths = new DisplayConfig.DisplayConfigPathInfo[pathCount];
        var modes = new DisplayConfig.DisplayConfigModeInfo[modeCount];

        result = DisplayConfig.QueryDisplayConfig(DisplayConfig.Qdc.OnlyActivePaths, ref pathCount, paths, ref modeCount, modes, IntPtr.Zero);
        if (result != DisplayConfig.ErrorSuccess)
        {
            return null;
        }

        DisplayConfig.DisplayConfigModeInfo primaryDisplayMode = modes
             .FirstOrDefault(IsPrimaryDisplayMode);

        DisplayConfig.DisplayConfigPathTargetInfo targetInfo =
            paths.FirstOrDefault(p => p.sourceInfo.id == primaryDisplayMode.id)
            .targetInfo;

        return targetInfo;
    }

    private static (bool?, bool?) GetHdrStatus(DisplayConfig.DisplayConfigPathTargetInfo targetInfo)
    {
        //Retrieve current HDR state
        var colorInfo = new DisplayConfig.DisplayConfigGetAdvancedColorInfo();
        colorInfo.header.type = DisplayConfig.DisplayConfigDeviceInfoType.GetAdvancedColorInfo;
        colorInfo.header.adapterId = targetInfo.adapterId;
        colorInfo.header.id = targetInfo.id;
        colorInfo.header.size = Marshal.SizeOf<DisplayConfig.DisplayConfigGetAdvancedColorInfo>();

        long result = DisplayConfig.DisplayConfigGetDeviceInfo(ref colorInfo.header);
        if (result != DisplayConfig.ErrorSuccess)
        {
            return (null, null);
        }

        return (colorInfo.advancedColorSupported, colorInfo.advancedColorEnabled);
    }
}